### PR TITLE
[core] Integrate mimalloc override allocators

### PR DESCRIPTION
**Pull request by espkk**
Initial mimalloc integration.
Tested for a long time with no issues, yet they may appear with some DLLs on some particular windows versions.

### CODE REVIEW

**Comment by ugeen4** [comment]
Tested. Don't see any problems with this PR !
 Why is the stable release (mimalloc 1.7.3) not being used?

**Comment by q4a** [comment]
Can you wait for 1-2 days before merge? I would like to test this changes on Linux.

**Comment by espkk** [comment]
@ugeen4, I tested both for quite a lot of time.
2.* is quite stable. the reason it is in beta I think is that it is not as well tested as 1.*
Considering the 2.* main difference is the internal resource management improvement which results in better performance, I decided to go with 2.* and see if it has some issues.
And I'm happy with it so far :)

@q4a, sure. let me know

**Review by ugeen4 — DISMISSED** [review]

File: src/apps/engine/src/compiler.cpp
Comment (q4a) [inline src/apps/engine/src/compiler.cpp]: Please, use `Segment.name.c_str()` to avoid warning/error:
```
error: cannot pass object of non-trivial type 'std::string' (aka 'basic_string<char>') through variadic method; call will abort at runtime [-Wnon-pod-varargs]
                     sizeof(uint32_t), Segment.name, fnsize);
                                               ^
```

**Comment by q4a** [comment]
src/apps/engine/src/compiler.cpp - line 3736.
Please, replace
`SetError("Segment (%s) not loaded", SegmentTable[segment_index].name);`
with
`SetError("Segment (%s) not loaded", SegmentTable[segment_index].name.c_str());`
to avoid warning-error:
```
error: cannot pass object of non-trivial type 'std::string' (aka 'basic_string<char>') through variadic method; call will abort at runtime [-Wnon-pod-varargs]
            SetError("Segment (%s) not loaded", SegmentTable[segment_index].name);
                                                                            ^
```

**Review by q4a — CHANGES_REQUESTED** [review]

File: src/apps/engine/CMakeLists.txt
Comment (q4a) [inline src/apps/engine/CMakeLists.txt]: Can you explain why you need to link mimalloc before c-runtime?
And if you really need it - try to avoid direct filenames: for example debug linux lib called `libmimalloc-debug.so`

**Review by q4a — CHANGES_REQUESTED** [review]

File: src/apps/engine/CMakeLists.txt
Comment (espkk) [reply src/apps/engine/CMakeLists.txt]: The idea is to override both C++ and C std memory handling. forC++ there is a well-defined way to override all variations of new/delete operators. 
However, for lower-level C alloc stuff it's not that simple. On windows, the allocator can only be injected at runtime.
This creates a constraint for linker/loader order. mimalloc-redirect shall be loaded strictly before C runtime (for modern windows it is ucrtbase/ucrtbased).

Regarding hardcoded names. I have no idea how we can deduce the names. I tried a few techniques and they do not work. We are heavily constrained by Conan API here.
One idea I have in mind - on Linux I think it can be statically linked. Therefore we can probably configure it at conanfile level depending on the platform. And disable the above code on Linux. But this is not something I can do myself right now. So feel free to either try it yourself or disable this code for your fork for now

File: src/apps/engine/src/compiler.cpp
Comment (espkk) [reply src/apps/engine/src/compiler.cpp]: good catch

File: src/apps/engine/CMakeLists.txt
Comment (espkk) [reply src/apps/engine/CMakeLists.txt:9]: good catch

Comment (q4a) [reply src/apps/engine/CMakeLists.txt]: I tested few vars like `${CONAN_LIBS_MIMALLOC_RELEASE}` and `${CONAN_PKG_LIBS_MIMALLOC_RELEASE}`, but they don't fit. So I can offer simple solution, which works fine on Windows and Linux
```
$ git diff
diff --git a/src/apps/engine/CMakeLists.txt b/src/apps/engine/CMakeLists.txt
index 1008563e..0a94b106 100644
--- a/src/apps/engine/CMakeLists.txt
+++ b/src/apps/engine/CMakeLists.txt
@@ -1,12 +1,18 @@
 include(${CMAKE_BINARY_DIR}/conanbuildinfo_multi.cmake)

+if (WIN32)
+set(MIMALLOC_DEP "$<IF:$<CONFIG:Release>,${CONAN_LIB_DIRS_MIMALLOC_RELEASE}/mimalloc.lib,${CONAN_LIB_DIRS_MIMALLOC_DEBUG}/mimalloc-debug.lib>")
+else()
+set(MIMALLOC_DEP "$<IF:$<CONFIG:Release>,${CONAN_LIB_DIRS_MIMALLOC_RELEASE}/libmimalloc.so,${CONAN_LIB_DIRS_MIMALLOC_DEBUG}/libmimalloc-debug.so>")
+endif()
+
 STORM_SETUP(
     TARGET_NAME engine
     TYPE executable
     DEPENDENCIES

     # this is something I really don't like but it seems there's not better way to ensure mimalloc is linked before c-runtime
-    $<IF:$<CONFIG:Release>,${CONAN_LIB_DIRS_MIMALLOC_RELEASE}/mimalloc.lib,${CONAN_LIB_DIRS_MIMALLOC_DEBUG}/mimalloc-debug.lib>
+    ${MIMALLOC_DEP}

     # internal
     common
```

Comment (espkk) [reply src/apps/engine/CMakeLists.txt]: I hope we'll find a better way... but let it be for now :)

**Review by q4a — APPROVED** [review]

**Comment by q4a** [comment]
Look good. Thanks.

**Review by ugeen4 — APPROVED** [review]

**Pull request merged**
